theano/tensor/nnet/spatialtf.py

```diff
@@ -0,0 +1,413 @@
+"""
+Concrete spatial transformer implementation
+"""
+from __future__ import absolute_import, print_function, division
+
+import theano
+from theano.gof.op import Op
+from theano.scalar import as_scalar
+from theano.tensor import as_tensor_variable
+from theano.tensor.extra_ops import cpu_contiguous
+from theano.gof import Apply
+from theano.gradient import grad_not_implemented
+import numpy as np
+
+
+def sampling_grid(height, width):
+    # Create sampling grid
+    x_t, y_t = np.meshgrid(np.linspace(-1, 1, width),
+                           np.linspace(-1, 1, height))
+    ones = np.ones(np.prod(x_t.shape))
+    grid = np.vstack([x_t.flatten(), y_t.flatten(), ones])
+    return grid
+
+
+class TransformerGrid(Op):
+    def make_node(self, theta, out_dims):
+        _theta = cpu_contiguous(as_tensor_variable(theta))
+
+        if _theta.ndim != 3:
+            raise TypeError('SpatialTransformerGrid (make_node) requires theta to '
+                            'be a 3D tensor; received "%s" (%i dims)' %
+                            (theta, _theta.ndim))
+
+        assert _theta.type.dtype in ('float16', 'float32', 'float64')
+
+        _out_dims = cpu_contiguous(as_tensor_variable(out_dims))
+        _out_dims = theano.tensor.basic.cast(out_dims, 'int64')
+
+        _grid = theano.tensor.tensor(dtype=_theta.type.dtype,
+                                     broadcastable=len(out_dims) * (False,))
+
+        return Apply(self, [_theta, _out_dims], [_grid])
+
+    def perform(self, node, inputs, output_storage):
+        theta, out_dims = inputs
+
+        assert len(out_dims) == 4
+        # Theta should be in the format (batch_size, 2, 3)
+        assert len(theta.shape) == 3
+        assert (theta.shape[1] == 2) and (theta.shape[2] == 3)
+        # Check if theta has the same batch size as out_dims
+        assert theta.shape[0] == out_dims[0]
+
+        num_batch = theta.shape[0]
+        grid_out = output_storage[0]
+
+        out_height, out_width = out_dims[2:]
+        grid = sampling_grid(out_height, out_width)
+        # Generate transformed grid with shape (num_batch, 2, out_height * out_width)
+        transformed_grid = np.dot(theta, grid)
+        # Dimshuffle grid into (2, num_batch, out_height * out_width)
+        transposed_grid = np.transpose(transformed_grid, axes=(1, 0, 2))
+        # Reshape into (2, num_batch, out_height, out_width)
+        grid_out[0] = np.reshape(transposed_grid, (2, num_batch, out_height, out_width)).astype(theta.dtype)
```

Comment:
It eases the CPU computations and (I assumed it) makes the grid data contiguous, but it can easily reverted to the same format as in the GPU. Otherwise, a dimshuffle is required when lifting to the GPU, as you've stated. I'll change it to use same format as the GPU.